### PR TITLE
overlord: start working on state engine along spec v2, have the main skeleton follow that

### PR DESCRIPTION
**Pull request by pedronis**
This starts working on state engine along spec v2, have the main skeleton follow that.

State and its tests are in an intermediate state, next branch will start from there with Lock and Unlock and proper checkpointing.

### CODE REVIEW

File: overlord/ifacemgr.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+// InterfaceManager is responsible for the maintenance of interfaces in
+// the system state.  It maintains interface connections, and also observes
+// installed snaps to track the current set of available plugs and slots.
+type InterfaceManager struct {
+	o *Overlord
+}
+
+// NewInterfaceManager returns a new InterfaceManager.
+func NewInterfaceManager(o *Overlord) (*InterfaceManager, error) {
```
Comment (niemeyer) [inline overlord/ifacemgr.go]: This should probably receive just what it really needs. The overlord is basically everything, which means everything will end up depending on everything.

File: overlord/overlord.go [hunk 2]
```diff
@@ -17,52 +17,52 @@
  *
  */
 
-// Package overlord implements the policies for state transitions for the operation of a snappy system.
+// Package overlord implements state changes for the operation of a snappy system.
```
Comment (niemeyer) [inline overlord/overlord.go]: s/state changes for the operation/the overall control/?

**Comment by niemeyer** [comment]
Thanks, LGTM!  I'd just drop that parameter from NewInterfaceManager at least.

File: overlord/ifacemgr.go (hunk 1 above)
Comment (pedronis) [inline overlord/ifacemgr.go]: we agreed to even go a bit further and split the managers into packages overlord/assertmgr overlord/snapmgr and overlord/ifacemgr

**Comment by mvo5** [comment]
Looks good, thanks! Once the issue that Gustavo raised is addressed this should go in.

**Pull request merged**